### ISSUE TITLE
Cluster Controller causes Rancher to OOM when too many resources are called from downstream

### ISSUE THREAD

**Issue by leflambeur** [rancher/rancher#26166]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
Create a downstream cluster and then 20k+ secrets in it

**Result:**
The cluster controller cannot start as the requests it makes to watch the k8s api take too long to complete - this causes a loop in the controller which results in an OOM kill

**Other details that may be helpful:**
Observed in 2.3.5

Scaling fixes in 2.4/Master would likely resolve in the short term:
https://github.com/rancher/rancher/blob/master/pkg/clustermanager/manager.go#L233

But not confident if this particular check is actual source of the timeout we're seeing

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):2.3.5
- Installation option (single install/HA):HA - RKE

**Downstream Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Imported kubeadm
- Machine type (cloud/VM/metal) and specifications (CPU/memory): 0.7 of 2k CPU 2.6 of 8TB RAM
- Kubernetes version (use `kubectl version`): 1.16.6

**Comment by dramich** [comment]
Some things to consider: Updating the timeout will help the cluster come in successfully. Once in though, trying to list the objects through the rancher API will still be slow, and this will vary depending on the item being requested as we process lists through our formatters/transformers. The listing and the timeout are both addressed in the 2.4 line.

**Comment by sowmyav27** [comment]
**Reproduced the issue on rancher:v2.3.5**

- Deploy a HA Rancher server
- Deploy a custom cluster - 1 etcd, 1 control plane and 3 worker nodes
- Create 10k+ secrets
- Delete all rancher pods
- User Cluster fails to come up - seen in `Unavailable` state.
- Rancher logs: 

```
2020/06/09 21:00:50 [INFO] Starting cluster controllers for c-d654n
2020/06/09 21:01:20 [INFO] Stopping cluster agent for c-d654n
E0609 21:01:20.070313       6 request.go:879] Unexpected error when reading response body: context deadline exceeded (Client.Timeout exceeded while reading body)
2020/06/09 21:01:23 [INFO] Registering project network policy
2020/06/09 21:01:23 [INFO] Registering CIS controller
2020/06/09 21:01:23 [INFO] registering podsecuritypolicy cluster handler for cluster c-d654n
2020/06/09 21:01:23 [INFO] registering podsecuritypolicy project handler for cluster c-d654n
2020/06/09 21:01:23 [INFO] registering podsecuritypolicy namespace handler for cluster c-d654n
2020/06/09 21:01:23 [INFO] registering podsecuritypolicy serviceaccount handler for cluster c-d654n
2020/06/09 21:01:23 [INFO] registering podsecuritypolicy template handler for cluster c-d654n
2020/06/09 21:01:23 [INFO] Registering monitoring for cluster "c-d654n"
2020/06/09 21:01:23 [INFO] Registering istio for cluster "c-d654n"
2020/06/09 21:01:23 [INFO] Starting cluster controllers for c-d654n
2020/06/09 21:01:53 [INFO] Stopping cluster agent for c-d654n
E0609 21:01:53.997149       6 request.go:879] Unexpected error when reading response body: context deadline exceeded (Client.Timeout exceeded while reading body)
2020/06/09 21:01:54 [ERROR] ClusterController _all_ [user-controllers-controller] failed with : failed to start user controllers for cluster c-d654n: timeout syncing controllers
2020/06/09 21:01:54 [INFO] Registering project network policy
2020/06/09 21:01:54 [INFO] Registering CIS controller
2020/06/09 21:01:54 [INFO] registering podsecuritypolicy cluster handler for cluster c-d654n
2020/06/09 21:01:54 [INFO] registering podsecuritypolicy project handler for cluster c-d654n
2020/06/09 21:01:54 [INFO] registering podsecuritypolicy namespace handler for cluster c-d654n
2020/06/09 21:01:54 [INFO] registering podsecuritypolicy serviceaccount handler for cluster c-d654n
2020/06/09 21:01:54 [INFO] registering podsecuritypolicy template handler for cluster c-d654n
2020/06/09 21:01:54 [INFO] Registering monitoring for cluster "c-d654n"
2020/06/09 21:01:54 [INFO] Registering istio for cluster "c-d654n"
2020/06/09 21:01:54 [INFO] Starting cluster controllers for c-d654n
2020/06/09 21:02:24 [INFO] Stopping cluster agent for c-d654n
E0609 21:02:24.060134       6 request.go:879] Unexpected error when reading response body: net/http: request canceled (Client.Timeout exceeded while reading body)
2020/06/09 21:02:29 [INFO] Registering project network policy
```

- Upgrade to **`2.4.5-rc2`**
- The user cluster comes up fine
- Rancher server logs:
```
I0609 21:27:27.760374       6 trace.go:116] Trace[651817040]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:26:57.760097191 +0000 UTC m=+139.885555423) (total time: 30.000232867s):
Trace[651817040]: [30.000232867s] [30.000232867s] END
E0609 21:27:27.760398       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.Service: Get https://<ip>:6443/api/v1/services?resourceVersion=39003&timeout=30s: context deadline exceeded
I0609 21:27:38.417460       6 trace.go:116] Trace[533333118]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:27:27.759352656 +0000 UTC m=+169.884810868) (total time: 10.658086403s):
Trace[533333118]: [10.658053668s] [10.658053668s] Objects listed
I0609 21:27:38.439520       6 trace.go:116] Trace[179198647]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:27:27.758230792 +0000 UTC m=+169.883688991) (total time: 10.681272814s):
Trace[179198647]: [10.681225421s] [10.681225421s] Objects listed
I0609 21:27:57.757537       6 trace.go:116] Trace[59751515]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:27:27.757214589 +0000 UTC m=+169.882672787) (total time: 30.000298192s):
Trace[59751515]: [30.000298192s] [30.000298192s] END
E0609 21:27:57.757555       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.APIService: Get https://<ip>:6443/apis/apiregistration.k8s.io/v1/apiservices?resourceVersion=39003&timeout=30s: context deadline exceeded
I0609 21:27:57.760667       6 trace.go:116] Trace[254283148]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:27:27.760408407 +0000 UTC m=+169.885866648) (total time: 30.000244656s):
Trace[254283148]: [30.000244656s] [30.000244656s] END
E0609 21:27:57.760676       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.Service: Get https://<ip>:6443/api/v1/services?resourceVersion=39003&timeout=30s: context deadline exceeded
I0609 21:28:27.761030       6 trace.go:116] Trace[22580427]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:27:57.760683833 +0000 UTC m=+199.886142020) (total time: 30.000326019s):
Trace[22580427]: [30.000326019s] [30.000326019s] END
E0609 21:28:27.761047       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.Service: Get https://<ip>:6443/api/v1/services?resourceVersion=39003&timeout=30s: context deadline exceeded
I0609 21:29:04.138043       6 trace.go:116] Trace[593412960]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:28:34.137663676 +0000 UTC m=+236.263121924) (total time: 30.000358356s):
Trace[593412960]: [30.000358356s] [30.000358356s] END
E0609 21:29:04.138057       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.APIService: Get https://<ip>:6443/apis/apiregistration.k8s.io/v1/apiservices?resourceVersion=39003&timeout=30s: context deadline exceeded
2020/06/09 21:29:37 [INFO] Creating new GlobalRoleBinding for GlobalRoleBinding grb-hrnqz
2020/06/09 21:29:37 [INFO] [mgmt-auth-grb-controller] Creating clusterRoleBinding for globalRoleBinding grb-hrnqz for user u-cvwmk with role cattle-globalrole-user
I0609 21:29:40.834523       6 trace.go:116] Trace[437300291]: "Reflector ListAndWatch" name:github.com/rancher/norman/controller/generic_controller.go:237 (started: 2020-06-09 21:29:10.834198128 +0000 UTC m=+272.959656373) (total time: 30.000304258s):
Trace[437300291]: [30.000304258s] [30.000304258s] END
E0609 21:29:40.834541       6 reflector.go:178] github.com/rancher/norman/controller/generic_controller.go:237: Failed to list *v1.Service: Get https://<ip>:6443/api/v1/services?resourceVersion=39003&timeout=30s: context deadline exceeded
```
- User is able to list 999 secrets out of 10k+. 
- API call to fetch the secrets first time takes about 2.8 minutes
- However there is a panic in the kube-apiserver of the user cluster - logged issue -https://github.com/rancher/rancher/issues/27519  to track this
